### PR TITLE
feat(tianmu): Support (Single) DELETE FROM tb_name and (Multiple) DELETE tb1_name,tb2_name FROM table_references (#343)

### LINKED ISSUE #344: feature: Support (Single-Table Syntax) DELETE FROM tb_name
**Is your feature request related to a problem? Please describe.**

>Delete the data of a single table and the whole table

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by konghaiya**
## Main task
- #343
## Subtask
- #344 
- #347

## Summary about this PR

Issue Number: close (#344) (#347)

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [x] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: mysql-test/suite/tianmu/r/delete.result [hunk 1]
```diff
@@ -0,0 +1,52 @@
+use test;
+CREATE TABLE `column_type_test` (
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/delete.result]: Pls, refer to innodb test cases at first for how to write a test case. 
If you DO NOT show the query log, you can disable the query log with `--disable_query_log
--enable_query_log` 
and to check whether the tianmu engine is ready or not, jus like innodb test case does. `--source include/have_innodb.inc`.
You can 
https://dev.mysql.com/doc/dev/mysql-server/latest/PAGE_MYSQLTEST_LANGUAGE_REFERENCE.html

File: mysql-test/suite/tianmu/r/delete.result [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+`c_mediumint` mediumint(9) DEFAULT NULL COMMENT 'mediumint',
+`c_int` int(11) DEFAULT NULL COMMENT 'int',
+`c_bigint` bigint(20) DEFAULT NULL COMMENT 'bigint',
+`c_float` float DEFAULT NULL COMMENT 'float',
+`c_double` double DEFAULT NULL COMMENT 'double',
+`c_decimal` decimal(10,5) DEFAULT NULL COMMENT 'decimal',
+`c_date` date DEFAULT NULL COMMENT 'date',
+`c_datetime` datetime DEFAULT NULL COMMENT 'datetime',
+`c_timestamp` timestamp NULL DEFAULT NULL COMMENT 'timestamp',
+`c_time` time DEFAULT NULL COMMENT 'time',
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/delete.result]: This test case IS NOT enough to ensure your code quality, need MORE test cases.

File: storage/tianmu/handler/tianmu_handler.cpp [hunk 3]
```diff
@@ -544,10 +544,7 @@ int TianmuHandler::delete_row([[maybe_unused]] const uchar *buf) {
  Called from sql_select.cc by JOIN::rein*it.
  Called from sql_union.cc by st_select_lex_unit::exec().
  */
-int TianmuHandler::delete_all_rows() {
```
Comment (RingsC) [inline storage/tianmu/handler/tianmu_handler.cpp:547]: Do we do access check(priviledges check)?

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/handler/tianmu_handler.cpp (hunk 3 above)
Comment (RingsC) [inline storage/tianmu/handler/tianmu_handler.cpp:547]: what will be happened under multi-sessions?

**Review by RingsC — CHANGES_REQUESTED** [review]

**Comment by hustjieke** [comment]
Add the usage of `Delete SQL syntax` to `StoneDB docs`.

File: storage/tianmu/handler/tianmu_handler.cpp (hunk 3 above)
Comment (konghaiya) [reply storage/tianmu/handler/tianmu_handler.cpp:547]: In the Tianmu engine, there is no need to judge the execution permission. The judgment of permission is already done at the beginning of the execution of the "execute" method in the SQL layer.
#### for example 1:
```c++
bool Sql_cmd_delete::execute(THD *thd)
{
  assert(thd->lex->sql_command == SQLCOM_DELETE);

  LEX *const lex= thd->lex;
  SELECT_LEX *const select_lex= lex->select_lex;
  SELECT_LEX_UNIT *const unit= lex->unit;
  TABLE_LIST *const first_table= select_lex->get_table_list();
  TABLE_LIST *const all_tables= first_table;

  if (delete_precheck(thd, all_tables))
    return true;
.............
```
#### for example 2:
```c++
bool Sql_cmd_delete_multi::execute(THD *thd)
{
  assert(thd->lex->sql_command == SQLCOM_DELETE_MULTI);

  bool res= false;
  LEX *const lex= thd->lex;
  SELECT_LEX *const select_lex= lex->select_lex;
  TABLE_LIST *const first_table= select_lex->get_table_list();
  TABLE_LIST *const all_tables= first_table;

  TABLE_LIST *aux_tables= thd->lex->auxiliary_table_list.first;
  uint del_table_count;
  Query_result_delete *del_result;

  if (multi_delete_precheck(thd, all_tables))
    return true;
........
```

#### for example 3:
```c++
bool Sql_cmd_truncate_table::execute(THD *thd)
{
  bool res= TRUE;
  TABLE_LIST *first_table= thd->lex->select_lex->table_list.first;
  DBUG_ENTER("Sql_cmd_truncate_table::execute");

  if (check_one_table_access(thd, DROP_ACL, first_table))
    DBUG_RETURN(res);
.......
```

> They all call check_ access() and check_ grant() judge the execution permission

File: mysql-test/suite/tianmu/r/delete.result (hunk 1 above)
Comment (konghaiya) [reply mysql-test/suite/tianmu/r/delete.result]: The test cases have been optimized

File: storage/tianmu/handler/tianmu_handler.cpp (hunk 3 above)
Comment (konghaiya) [reply storage/tianmu/handler/tianmu_handler.cpp:547]: In the SQL layer, the tables to be operated have been locked to support users' synchronous and mutually exclusive access. There is no need to perform parallel access control again in the engine.

File: storage/tianmu/handler/ha_rcengine.cpp [hunk 4]
```diff
@@ -144,5 +144,47 @@ bool tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void *arg)
   return false;
 }
 
+bool TianMuDeleteAllMulti(THD *thd, TABLE_LIST *table_list, ha_rows &deleted){
```
Comment (hustjieke) [inline storage/tianmu/handler/ha_rcengine.cpp]: `TianMuDeleteAllMulti`--->`TianmuDeleteMultiTbl`

File: storage/tianmu/handler/ha_rcengine.cpp [hunk 5]
```diff
@@ -144,5 +144,47 @@ bool tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void *arg)
...
+
+  if (!core::Engine::IsTIANMUTable(table_list->table)) return true;
```
Comment (hustjieke) [inline storage/tianmu/handler/ha_rcengine.cpp]: `IsTIANMUTable`--->`IsTianmuTbl`

File: sql/sql_delete.cc [hunk 6]
```diff
@@ -1445,6 +1448,15 @@ bool Sql_cmd_delete_multi::execute(THD *thd)
     return true;
   }
 
+  // Tianmu engine multi table data clearing
+  if (!(res = Tianmu::dbhandler::TianmuDeleteMultiTbl(thd, all_tables,
```
Comment (andyli029) [inline sql/sql_delete.cc:1452]: The better way？

**Comment by hustjieke** [comment]
Missing docs:
1. Single delete syntax.
2. How to use.

File: mysql-test/suite/tianmu/t/multi_table_delete.test [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+values(105,105,105,105,105,5.2,10.88,105.083,'2016-02-25','2016-02-25 10:20:01','2016-02-25 05:20:01','10:20:01','stoneatom','hello','bcdefghijklmn');
+select * from t1;
+select count(*) from t1;
+select max(c_int) from t1;
+select min(c_int) from t1;
+select avg(c_int) from t1;
+select sum(c_int) from t1;
+
+insert into t2(c_tinyint,c_smallint,c_mediumint,c_int,c_bigint,c_float,c_double,c_decimal,c_date,c_datetime,c_timestamp,c_time,c_char,c_varchar,c_text)
+values(105,105,105,105,105,5.2,10.88,105.083,'2016-02-25','2016-02-25 10:20:01','2016-02-25 05:20:01','10:20:01','stoneatom','hello','bcdefghijklmn');
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/multi_table_delete.test:93]: Adds some test cases for unicode. such as  insert unicode chars, in values, table name, or etc.

File: sql/sql_delete.cc (hunk 6 above)
Comment (RingsC) [reply sql/sql_delete.cc:1452]: > The better way？
+1

File: storage/tianmu/core/engine.h [hunk 8]
```diff
@@ -145,7 +145,7 @@ class Engine final {
   static AttributeTypeInfo GetCorrespondingATI(Field &field);
   static AttributeTypeInfo GetAttrTypeInfo(const Field &field);
   static common::CT GetCorrespondingType(const enum_field_types &eft);
-  static bool IsTIANMUTable(TABLE *table);
+  static bool IsTianmuTbl(TABLE *table);
```
Comment (RingsC) [inline storage/tianmu/core/engine.h:148]: `XXXTable` is better than `XXXTbl`.

File: storage/tianmu/handler/ha_rcengine.cpp [hunk 9]
```diff
@@ -144,5 +144,47 @@ bool tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void *arg)
   return false;
 }
 
+bool TianmuDeleteMultiTbl(THD *thd, TABLE_LIST *table_list, ha_rows &deleted){
+
+  if (!core::Engine::IsTianmuTbl(table_list->table)) return true;
```
Comment (RingsC) [inline storage/tianmu/handler/ha_rcengine.cpp:149]: Adding fault injection codes, just like MySQL does. What will happens that if the instance crash in deleting operation?

Comment (RingsC) [reply storage/tianmu/handler/ha_rcengine.cpp:149]: And how about deleting rows from a temperoary table?

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/handler/ha_rcengine.cpp [hunk 10]
```diff
@@ -144,5 +144,47 @@ bool tianmu_load(THD *thd, sql_exchange *ex, TABLE_LIST *table_list, void *arg)
...
+      TABLE_LIST* delete_table_ref;
+      for (delete_table_ref = table_list; delete_table_ref; delete_table_ref = delete_table_ref->next_local)
+      {
+          TABLE *const table= delete_table_ref->table;
+          if (table && table->file)
+          {
+              // Update the table->file->stats.records number
+              table->file->info(HA_STATUS_VARIABLE | HA_STATUS_NO_LOCK);
+              ha_rows const maybe_deleted= table->file->stats.records;
+              if(!table->file->ha_truncate())
```
Comment (RingsC) [inline storage/tianmu/handler/ha_rcengine.cpp:176]: `TianmuDeleteMultiTbl` means delete some rows not whole table rows, but in your function body just do `truncate` operation. misleading function name.

**Pull request closed without merging**